src/social.ts

```diff
@@ -0,0 +1,177 @@
+import { ArcadeIdentity } from "./ident";
+import { NostrPool } from "./pool";
+import { EventTemplate, Filter } from "nostr-tools";
+
+const STALE_GRAPH = 1000 * 60 * 60 * 24 * 7 // 1 week
+const GRAPH_DEPTH = 3
+
+// define a nip02 tag type
+type NIP02Contact = string[];
+type PublicKey = string;
+type SocialGraphEntry = {
+  pubkey: PublicKey,
+  degree: number,
+  connection: PublicKey, // For degree 1, this is always the user's own pubkey. For degree 2, this is the pubkey of the contact between us and this contactPubkey.
+  // meta: {},
+  lastUpdated: number, // timestamp is updated when this contact's social graph is updated
+  fwf?: number, // (number of) friends who follow (this contact)
+}
+type SocialGraph = {
+  [key in PublicKey]: SocialGraphEntry;
+}
+
+/**
+ * Ensure the provided ptag is a valid NIP02 contact we can work with.
+ * @param ptag string[p, pubkey, relay?, pet name?]
+ * @returns 
+ */
+export function isValidNIP02Contact(ptag: NIP02Contact): ptag is NIP02Contact {
+  return Array.isArray(ptag) &&
+    ptag.length >= 2 &&
+    ptag.length <= 4 &&
+    ptag.every(item => typeof item === 'string') &&
+    ptag[0] === 'p' &&
+    ptag[1].length === 64
+}
+
+export class ArcadeSocial {
+  public pool: NostrPool;
+  private ident: ArcadeIdentity;
+  public socialGraph: SocialGraph = {};
+  public iteration = 0;
+  constructor(pool: NostrPool, ident: ArcadeIdentity) {
+    this.pool = pool;
+    this.ident = ident;
+    this.generate(ident);
+  }
+  /**
+   * Kick off social graph generation process
+   * @param ident ArcadeIdentity
+   */
+  generate(ident: ArcadeIdentity): void {
+    this.extendGraph(ident.pubKey);
+  }
+  /**
+   * Provide a public key to extend the social graph by one degree. Hey that rhymes!
+   * @param pubkey PublicKey - raw hex public key
+   * @param degree the current degree of separation from the user in this iteration
+   */
+  extendGraph(pubkey: PublicKey, degree = 1): void {
+    const kind3: EventTemplate[] = [];
+    const filter: Filter<number> = { kinds: [3], authors: [pubkey] };
+    this.pool.sub([filter], event => kind3.push(event), () => {
```

Comment:
you're gonna need kind0 as well.   most clients stick the contacts in tags in kind0 instead.

Reply:
I've been having trouble finding an example of this. Do they put the contacts in the tags the same way as kind 3, but just in a kind 0 event?